spectral_cube/dask_spectral_cube.py

```diff
@@ -0,0 +1,1278 @@
+            coordinates
+        spatial_only: bool
+            Return only slices that affect the spatial dimensions; the spectral
+            dimension will be left unchanged
+        """
+
+        # We need to use a slightly different approach to SpectralCube here
+        # because there isn't yet a dask-friendly version of find_objects
+        # https://github.com/dask/dask-image/issues/96
+
+        if isinstance(region_mask, np.ndarray):
+            if is_broadcastable_and_smaller(region_mask.shape, self.shape):
+                region_mask = BooleanArrayMask(region_mask, self._wcs)
+            else:
+                raise ValueError("Mask shape does not match cube shape.")
+
+        include = region_mask.include(self._data, self._wcs,
+                                      wcs_tolerance=self._wcs_tolerance)
+
+        include = da.broadcast_to(include, self.shape)
+
+        slices = []
+        for axis in range(3):
+            collapse_axes = tuple(index for index in range(3) if index != axis)
+            valid = self._compute(da.any(include, axis=collapse_axes))
+            if np.any(valid):
+                indices = np.where(valid)[0]
+                slices.append(slice(np.min(indices), np.max(indices) + 1))
+            else:
+                slices.append(slice(0))
+
+        if spatial_only:
+            slices = (slice(None), slices[1], slices[2])
+
+        return tuple(slices)
+
+    def downsample_axis(self, factor, axis, estimator=np.nanmean,
+                        truncate=False):
+        """
+        Downsample the cube by averaging over *factor* pixels along an axis.
+        Crops right side if the shape is not a multiple of factor.
+
+        The WCS will be 'downsampled' by the specified factor as well.
+        If the downsample factor is odd, there will be an offset in the WCS.
+
+        There is both an in-memory and a memory-mapped implementation; the
+        default is to use the memory-mapped version.  Technically, the 'large
+        data' warning doesn't apply when using the memory-mapped version, but
+        the warning is still there anyway.
+
+        Parameters
+        ----------
+        myarr : `~numpy.ndarray`
+            The array to downsample
+        factor : int
+            The factor to downsample by
+        axis : int
+            The axis to downsample along
+        estimator : function
+            defaults to mean.  You can downsample by summing or
+            something else if you want a different estimator
+            (e.g., downsampling error: you want to sum & divide by sqrt(n))
+        truncate : bool
+            Whether to truncate the last chunk or average over a smaller number.
+            e.g., if you downsample [1,2,3,4] by a factor of 3, you could get either
+            [2] or [2,4] if truncate is True or False, respectively.
+        """
+
+        # FIXME: this does not work correctly currently due to
+        # https://github.com/dask/dask/issues/6102
```

Comment:
I added a warning:

```
        warnings.warn('In some cases, the final shape of the output from downsample_axis '
                      'is incorrect, so use the result with caution', UserWarning)
```

I thought it's still worth keeping since it kind of works but might have off-by-one errors.